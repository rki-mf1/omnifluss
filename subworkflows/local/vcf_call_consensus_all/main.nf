include { BCFTOOLS_FILTER                            } from '../../../modules/nf-core/bcftools/filter/main'
include { INV_GET_DELETIONS_PYVCF                    } from '../../../modules/local/inv_get_deletions_pyvcf/main'
include { INV_SET_GT_BCFTOOLS                        } from '../../../modules/local/inv_set_gt_bcftools/main'
include { CREATE_MASK_CONSENSUS                      } from '../../../modules/local/inv_consensus_bedtools/main'
include { TABIX_TABIX                                } from '../../../modules/nf-core/tabix/tabix/main'
include { BCFTOOLS_CONSENSUS                         } from '../../../modules/nf-core/bcftools/consensus/main'

workflow VCF_CALL_CONSENSUS_ALL {
    take:
    tools                       // string
    val_consensus_mincov        // integer
    ch_ref                      // channel: [ val(meta), fasta ]
    ch_vcf                      // channel: [ val(meta), vcf   ]
    ch_bam                      // channel: [ val(meta), bam   ]
    ch_rescued_variants         // channel: [ val(meta), bed   ]

    main:
    ch_versions                 = Channel.empty()

    if (tools.split(',').contains('bcftools')) {

        BCFTOOLS_FILTER(
            ch_vcf                      // channel: [ val(meta), vcf ]
        )
        .vcf
        | set {ch_filtered_vcf}

        ch_versions = ch_versions.mix(BCFTOOLS_FILTER.out.versions.first())

        // mask
        INV_GET_DELETIONS_PYVCF(
            ch_filtered_vcf             // channel: [ val(meta), vcf ]
        )
        .del_vcf
        | set {ch_del_adjusted_vcf}
        //note: no version save, because bcftools is used and it was incorporated prior

        // comprised createMaskConsensus & createMaskConsensus_special_variant_case in this module
        CREATE_MASK_CONSENSUS(
            val_consensus_mincov,       // integer
            ch_del_adjusted_vcf,        // channel: [ val(meta), vcf ]
            ch_bam,                     // channel: [ val(meta), bam ]
            ch_rescued_variants         // channel: [ val(meta), bed ]
        ).final_bed
        | set {ch_final_bed}

        ch_versions = ch_versions.mix(CREATE_MASK_CONSENSUS.out.versions.first())

        // vcf
        INV_SET_GT_BCFTOOLS(
            ch_filtered_vcf             // channel: [ val(meta), vcf ]
        ).vcf
        | set {ch_final_vcf}

        ch_versions = ch_versions.mix(INV_SET_GT_BCFTOOLS.out.versions.first())

        TABIX_TABIX(
            ch_final_vcf                // channel: [ val(meta), vcf ]
        ).tbi
        | set {ch_final_vcf_tbi}

        ch_versions = ch_versions.mix(TABIX_TABIX.out.versions.first())

        consensus_input = ch_final_vcf.join(ch_final_vcf_tbi, by:[0])
        consensus_input = consensus_input.combine( ch_ref.map{ it[1] } )
        consensus_input = consensus_input.join(ch_final_bed, by:[0])

        BCFTOOLS_CONSENSUS(
            consensus_input // channel: [ val(meta), vcf, tbi, fasta, bed ]
        )
        //note: no version save, because bcftools is used and it was incorporated prior
    }

    emit:
    consensus_calls     = BCFTOOLS_CONSENSUS.out.fasta
    versions            = ch_versions


}
