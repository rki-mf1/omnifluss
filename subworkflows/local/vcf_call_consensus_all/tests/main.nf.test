nextflow_workflow {

    name "Test Subworkflow VCF_CALL_CONSENSUS_ALL"
    script "../main.nf"
    config "./nextflow.config"
    workflow "VCF_CALL_CONSENSUS_ALL"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/vcf_call_consensus_all"
    tag "bedtools"
    tag "bcftools/filter"
    tag "tabix/tabix"
    tag "bcftools/consensus"
    tag "pyvcf"
    tag "inv_create_consensus_mask_bedtools"
    tag "inv_set_gt_bcftools"


    test("vcf_call_consensus_all") {
        options '-stub'
        when {
            workflow {
                """
                input[0] = "bcftools"
                input[1] = 10
                input[2] = Channel.value([
                    [ id: "ref", single_end:false],
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true) ]
                ])
                input[3] = Channel.of([
                    [ id:'test', single_end:false],
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true) ]
                ]  )
                input[4] = Channel.of([
                    [ id:'test', single_end:false],
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.bam', checkIfExists: true) ]
                ])
                input[5] = Channel.of([
                    [ id:'test', single_end:false],
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bedgraph/test.bedgraph', checkIfExists: true) ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
