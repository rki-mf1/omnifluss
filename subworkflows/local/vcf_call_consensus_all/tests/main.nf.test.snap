{
    "vcf_call_consensus_all": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,0686151f8904a178e94d7b45febf55fa",
                    "versions.yml:md5,4b863f593f80aaf04f25682c02c04049",
                    "versions.yml:md5,53e6daebdd619151ad85e86f809a4ccc",
                    "versions.yml:md5,d90f50b56cb7bff2ecda7ef3bba5d74a",
                    "versions.yml:md5,ee1d9a95e761a2cf1b6505abe9be44de"
                ],
                "consensus_calls": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0686151f8904a178e94d7b45febf55fa",
                    "versions.yml:md5,4b863f593f80aaf04f25682c02c04049",
                    "versions.yml:md5,53e6daebdd619151ad85e86f809a4ccc",
                    "versions.yml:md5,d90f50b56cb7bff2ecda7ef3bba5d74a",
                    "versions.yml:md5,ee1d9a95e761a2cf1b6505abe9be44de"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.1"
        },
        "timestamp": "2024-07-31T11:44:16.242615092"
    }
}